// @sap/ux-ui5-tooling/dist/preview-middleware/dist/client/adp/init.js
"use strict";

sap.ui.define(["sap/base/Log", "open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common", "../utils/version", "../cpe/init", "./sync-views-utils", "../utils/application", "./quick-actions/load", "./init-dialogs", "../utils/info-center-message", "../cpe/communication-service", "./change-file-validator"], function (log, ___sap_ux_private_control_property_editor_common, ___utils_versionjs, __init, ___sync_views_utilsjs, ___utils_applicationjs, ___quick_actions_loadjs, ___init_dialogsjs, ___utils_info_center_messagejs, ___cpe_communication_servicejs, ___change_file_validatorjs) {
  "use strict";

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule && typeof obj.default !== "undefined" ? obj.default : obj;
  }
  function __ui5_require_async(path) {
    return new Promise(function (resolve, reject) {
      sap.ui.require([path], function (module) {
        if (!(module && module.__esModule)) {
          module = module === null || !(typeof module === "object" && path.endsWith("/library")) ? {
            default: module
          } : module;
          Object.defineProperty(module, "__esModule", {
            value: true
          });
        }
        resolve(module);
      }, function (err) {
        reject(err);
      });
    });
  }
  const enableTelemetry = ___sap_ux_private_control_property_editor_common["enableTelemetry"];
  const MessageBarType = ___sap_ux_private_control_property_editor_common["MessageBarType"];
  const toggleAppPreviewVisibility = ___sap_ux_private_control_property_editor_common["toggleAppPreviewVisibility"];
  const getFullyQualifiedUi5Version = ___utils_versionjs["getFullyQualifiedUi5Version"];
  const getUi5Version = ___utils_versionjs["getUi5Version"];
  const isLowerThanMinimalUi5Version = ___utils_versionjs["isLowerThanMinimalUi5Version"];
  const minVersionInfo = ___utils_versionjs["minVersionInfo"];
  const init = _interopRequireDefault(__init);
  const updateSyncViewsIds = ___sync_views_utilsjs["updateSyncViewsIds"];
  const showSyncViewsWarning = ___sync_views_utilsjs["showSyncViewsWarning"];
  const getApplicationType = ___utils_applicationjs["getApplicationType"];
  const loadDefinitions = ___quick_actions_loadjs["loadDefinitions"];
  const initDialogs = ___init_dialogsjs["initDialogs"];
  const sendInfoCenterMessage = ___utils_info_center_messagejs["sendInfoCenterMessage"];
  const CommunicationService = ___cpe_communication_servicejs["CommunicationService"];
  const initOrphanedChangeDetection = ___change_file_validatorjs["initOrphanedChangeDetection"];
  var __exports = async function (rta) {
    const flexSettings = rta.getFlexSettings();
    if (flexSettings.telemetry === true) {
      enableTelemetry();
    }
    const ui5VersionInfo = await getUi5Version();

    // Plugins need to be set before adding additional plugins to prevent overriding with the default
    // and allow usage of getPlugins later in the flow
    const defaultPlugins = rta.getDefaultPlugins();
    rta.setPlugins(defaultPlugins);
    if (isLowerThanMinimalUi5Version(ui5VersionInfo, {
      major: 1,
      minor: 136,
      patch: 2
    })) {
      await initDialogs(rta, ui5VersionInfo);
    } else {
      (await __ui5_require_async('open/ux/preview/client/adp/add-fragment')).initAddXMLPlugin(rta);
      (await __ui5_require_async('open/ux/preview/client/adp/extend-controller')).initExtendControllerPlugin(rta);
    }
    if (!isLowerThanMinimalUi5Version(ui5VersionInfo, {
      major: 1,
      minor: 78
    })) {
      const ExtensionPointService = (await __ui5_require_async('open/ux/preview/client/adp/extension-point')).default;
      const extPointService = new ExtensionPointService(rta);
      extPointService.init();
    }
    const manifest = rta.getRootControlInstance().getManifest();
    const applicationType = getApplicationType(manifest);
    const quickActionRegistries = await loadDefinitions(applicationType);

    // Initialize OVP bridge functions if the application is an OVP app
    if (manifest['sap.ovp']) {
      const {
        initOvpWindowFunctions
      } = await __ui5_require_async('open/ux/preview/client/adp/ovp-window-functions');
      initOvpWindowFunctions();
    }
    await init(rta, quickActionRegistries);

    // Register synchronious views detection and warning
    // This is not awaited to prevent deadlock in the initialization
    rta.getService('outline').then(outlineService => {
      outlineService.attachEvent('update', async () => {
        await updateSyncViewsIds(ui5VersionInfo);
        await showSyncViewsWarning();
      });
    }).catch(error => {
      log.error('Failed to attach update event to outline service', error);
    });
    if (isLowerThanMinimalUi5Version(ui5VersionInfo)) {
      await sendInfoCenterMessage({
        title: {
          key: 'FLP_UI5_VERSION_WARNING_TITLE'
        },
        description: {
          key: 'FLP_UI5_VERSION_WARNING_DESCRIPTION',
          params: [getFullyQualifiedUi5Version(ui5VersionInfo), getFullyQualifiedUi5Version(minVersionInfo)]
        },
        type: MessageBarType.error
      });
      CommunicationService.sendAction(toggleAppPreviewVisibility(false));
      return;
    }
    initOrphanedChangeDetection().catch(error => {
      log.error('Failed to run orphaned change detection', error);
    });
    log.debug('ADP init executed.');
  };
  return __exports;
});
