// @sap/ux-ui5-tooling/dist/preview-middleware/dist/client/adp/quick-actions/common/create-page-action.js
"use strict";

sap.ui.define(["sap/ui/dt/OverlayRegistry", "../../dialog-factory", "../simple-quick-action-base", "../../../utils/application", "../../../utils/version", "../dialog-enablement-validator"], function (OverlayRegistry, ____dialog_factoryjs, ___simple_quick_action_basejs, _____utils_applicationjs, _____utils_versionjs, ___dialog_enablement_validatorjs) {
  "use strict";

  const DialogFactory = ____dialog_factoryjs["DialogFactory"];
  const DialogNames = ____dialog_factoryjs["DialogNames"];
  const SimpleQuickActionDefinitionBase = ___simple_quick_action_basejs["SimpleQuickActionDefinitionBase"];
  const getApplicationType = _____utils_applicationjs["getApplicationType"];
  const getUi5Version = _____utils_versionjs["getUi5Version"];
  const isLowerThanMinimalUi5Version = _____utils_versionjs["isLowerThanMinimalUi5Version"];
  const DIALOG_ENABLEMENT_VALIDATOR = ___dialog_enablement_validatorjs["DIALOG_ENABLEMENT_VALIDATOR"];
  const ADD_PAGE_ACTION = 'add-page-action';
  const CONTROL_TYPES = ['sap.f.DynamicPageTitle', 'sap.uxap.ObjectPageHeader', 'sap.uxap.ObjectPageDynamicHeaderTitle'];

  /**
   * Quick Action for adding a custom page action.
   */
  class AddPageActionQuickAction extends SimpleQuickActionDefinitionBase {
    constructor(context) {
      super(ADD_PAGE_ACTION, CONTROL_TYPES, 'QUICK_ACTION_ADD_CUSTOM_PAGE_ACTION', context, [DIALOG_ENABLEMENT_VALIDATOR]);
      this.appType = getApplicationType(this.context.rta.getRootControlInstance().getManifest());
    }
    async initialize() {
      const version = await getUi5Version();
      if (this.appType === 'fe-v4' && isLowerThanMinimalUi5Version(version, {
        major: 1,
        minor: 130
      })) {
        return;
      }
      await super.initialize();
    }
    async execute() {
      if (this.control) {
        const overlay = OverlayRegistry.getOverlay(this.control) || [];
        await DialogFactory.createDialog(overlay, this.context.rta, DialogNames.ADD_FRAGMENT, undefined, {
          aggregation: 'actions',
          title: 'QUICK_ACTION_ADD_CUSTOM_PAGE_ACTION',
          defaultAggregationArrayIndex: 1
        }, {
          actionName: this.type,
          telemetryEventIdentifier: this.getTelemetryIdentifier()
        });
      }
      return [];
    }
  }
  var __exports = {
    __esModule: true
  };
  __exports.ADD_PAGE_ACTION = ADD_PAGE_ACTION;
  __exports.AddPageActionQuickAction = AddPageActionQuickAction;
  return __exports;
});
