// @sap/ux-ui5-tooling/dist/preview-middleware/dist/client/cpe/utils.js
"use strict";

sap.ui.define(["sap/ui/core/mvc/XMLView", "sap/ui/core/UIComponent", "../utils/core"], function (XMLView, UIComponent, ___utils_corejs) {
  "use strict";

  function __ui5_require_async(path) {
    return new Promise(function (resolve, reject) {
      sap.ui.require([path], function (module) {
        if (!(module && module.__esModule)) {
          module = module === null || !(typeof module === "object" && path.endsWith("/library")) ? {
            default: module
          } : module;
          Object.defineProperty(module, "__esModule", {
            value: true
          });
        }
        resolve(module);
      }, function (err) {
        reject(err);
      });
    });
  }
  const getComponent = ___utils_corejs["getComponent"];
  /**
   * Get runtime control.
   *
   * @param overlayControl - element overlay.
   * @returns ManagedObject
   */
  function getRuntimeControl(overlayControl) {
    let runtimeControl;
    if (overlayControl.getElementInstance) {
      runtimeControl = overlayControl.getElementInstance();
    } else {
      runtimeControl = overlayControl.getElement();
    }
    return runtimeControl;
  }

  /**
   * Get library of a control name.
   *
   * @param controlName - name of the ui5 control eg: sap.m.Button.
   * @returns Promise<string>
   */
  async function getLibrary(controlName) {
    return new Promise(resolve => {
      const controlPath = controlName.replace(/\./g, '/');
      sap.ui.require([controlPath], control => {
        const contMetadata = control.getMetadata();
        // getLibraryName method does not exist on events
        if (contMetadata?.getLibraryName) {
          const contLibName = contMetadata.getLibraryName();
          resolve(contLibName);
        } else {
          resolve(''); // return empty for events
        }
      });
    });
  }

  /**
   * Gets the root view of component for the provided ComponentContainer control.
   *
   * @param container ComponentContainer control.
   * @returns XMLView which is the root control of the component if it exists.
   */
  function getRootControlFromComponentContainer(container) {
    if (container) {
      const componentId = container.getComponent();
      const component = getComponent(componentId);
      if (component instanceof UIComponent) {
        const rootControl = component.getRootControl();
        if (rootControl instanceof XMLView) {
          return rootControl;
        }
      }
    }
    return undefined;
  }
  function getManifestProperties(control, changeService, controlOverlay) {
    const overlayData = controlOverlay?.getDesignTimeMetadata().getData();
    if (!controlOverlay || !overlayData?.manifestSettings) {
      return {};
    }
    const manifestPropertiesValue = overlayData?.manifestSettingsValues(overlayData?.manifestSettings(control), control);
    const manifestProperties = overlayData?.manifestSettings(control).reduce((acc, item) => {
      const propertyId = item.path ?? item.id;
      const value = changeService.getConfigurationPropertyValue(control.getId(), propertyId);
      let propertyValue = value === 0 || value === false || value ? value : manifestPropertiesValue[propertyId];
      if (item?.type && ['boolean', 'number', 'string'].includes(item?.type)) {
        if (propertyValue === undefined) {
          propertyValue = item.value; // set default value of property
        }
      }
      if (!acc[propertyId]) {
        acc[propertyId] = {
          ...item,
          defaultValue: item.value,
          configuration: true,
          name: item.id,
          readableName: item.name ?? item.id,
          manifestPropertyPath: `${overlayData?.manifestPropertyPath(control)}/${propertyId}`,
          type: item.type === 'number' ? 'int' : item.type,
          value: propertyValue
        };
      }
      return acc;
    }, {});
    return manifestProperties;
  }

  // eslint-disable-next-line @sap-ux/fiori-tools/sap-no-global-variable
  let overlayModulesPromise;

  /**
   * Lazily loads and caches the OverlayRegistry and OverlayUtil modules.
   * The dynamic imports are executed only once; subsequent calls return the same resolved promise.
   *
   * @returns Promise resolving to an object containing the OverlayRegistry and OverlayUtil modules.
   */
  function getOverlayModules() {
    overlayModulesPromise ??= Promise.all([__ui5_require_async('sap/ui/dt/OverlayRegistry'), __ui5_require_async('sap/ui/dt/OverlayUtil')]).then(([{
      default: OverlayRegistry
    }, {
      default: OverlayUtil
    }]) => ({
      OverlayRegistry,
      OverlayUtil
    }));
    return overlayModulesPromise;
  }
  const getOverlay = async control => {
    const {
      OverlayRegistry,
      OverlayUtil
    } = await getOverlayModules();
    let controlOverlay = OverlayRegistry.getOverlay(control);
    if (!controlOverlay?.getDomRef()) {
      //look for closest control
      controlOverlay = OverlayUtil.getClosestOverlayFor(control);
    }
    return controlOverlay;
  };
  var __exports = {
    __esModule: true
  };
  __exports.getRuntimeControl = getRuntimeControl;
  __exports.getLibrary = getLibrary;
  __exports.getRootControlFromComponentContainer = getRootControlFromComponentContainer;
  __exports.getManifestProperties = getManifestProperties;
  __exports.getOverlay = getOverlay;
  return __exports;
});
