// @sap/ux-ui5-tooling/dist/preview-middleware/dist/client/thirdparty/@sap-ux-private/control-property-editor-common.js
sap.ui.define(['exports'], (function (exports) { 'use strict';

    const BOOLEAN_VALUE_TYPE = 'boolean';
    const INTEGER_VALUE_TYPE = 'integer';
    const FLOAT_VALUE_TYPE = 'float';
    const STRING_VALUE_TYPE = 'string';
    const INPUT_EDITOR_TYPE = 'input';
    const DROPDOWN_EDITOR_TYPE = 'dropdown';
    const CHECKBOX_EDITOR_TYPE = 'checkbox';
    const SCENARIO = {
        AppVariant: 'APP_VARIANT',
        VersionedAppVariant: 'VERSIONED_APP_VARIANT',
        AdaptationProject: 'ADAPTATION_PROJECT',
        FioriElementsFromScratch: 'FE_FROM_SCRATCH',
        UiAdaptation: 'UI_ADAPTATION'
    };
    exports.PropertyType = void 0;
    (function (PropertyType) {
        PropertyType["Configuration"] = "configuration";
        PropertyType["ControlProperty"] = "controlProperty";
    })(exports.PropertyType || (exports.PropertyType = {}));
    const PENDING_CHANGE_TYPE = 'pending';
    const SAVED_CHANGE_TYPE = 'saved';
    const UNKNOWN_CHANGE_KIND = 'unknown';
    const GENERIC_CHANGE_KIND = 'generic';
    const CONTROL_CHANGE_KIND = 'control';
    const SIMPLE_QUICK_ACTION_KIND = 'simple';
    const NESTED_QUICK_ACTION_KIND = 'nested';
    exports.MessageBarType = void 0;
    (function (MessageBarType) {
        /** Info styled MessageBar */
        MessageBarType[MessageBarType["info"] = 0] = "info";
        /** Error styled MessageBar */
        MessageBarType[MessageBarType["error"] = 1] = "error";
        /** Warning styled MessageBar */
        MessageBarType[MessageBarType["warning"] = 5] = "warning";
    })(exports.MessageBarType || (exports.MessageBarType = {}));
    /**
     * Create matcher.
     *
     * @param type action type
     * @returns (value: { type: unknown } | undefined) => value is Y
     */
    function createMatcher(type) {
        return function match(value) {
            return value?.type === type;
        };
    }
    /**
     * Create action factory.
     *
     * @param prefix to determine ext action
     * @returns Function
     */
    function createActionFactory(prefix) {
        return function createAction(name) {
            const actionType = [prefix, name].join(' ');
            /**
             *
             * @param payload action payload
             * @returns PayloadAction<typeof actionType, T>
             */
            function action(payload) {
                return {
                    type: actionType,
                    payload
                };
            }
            action.type = actionType;
            action.match = createMatcher(actionType);
            return action;
        };
    }
    const PENDING_SUFFIX = '<pending>';
    const FULFILLED_SUFFIX = '<fulfilled>';
    const REJECTED_SUFFIX = '<rejected>';
    /**
     * Factory for creating  request response actions.
     *
     * @param prefix action prefix
     * @returns Function
     */
    function createAsyncActionFactory(prefix) {
        return function createAction(name) {
            const pendingType = [prefix, name, PENDING_SUFFIX].join(' ');
            /**
             * Pending action.
             *
             * @param payload action payload
             * @returns PayloadAction<typeof pendingType, T>
             */
            function pending(payload) {
                return {
                    type: pendingType,
                    payload
                };
            }
            pending.type = pendingType;
            pending.match = createMatcher(pendingType);
            const fulfilledType = [prefix, name, FULFILLED_SUFFIX].join(' ');
            /**
             * Fulfill action.
             *
             * @param payload action payload
             * @returns PayloadAction<typeof fulfilledType, F>
             */
            function fulfilled(payload) {
                return {
                    type: fulfilledType,
                    payload
                };
            }
            fulfilled.type = fulfilledType;
            fulfilled.match = createMatcher(fulfilledType);
            const rejectedType = [prefix, name, REJECTED_SUFFIX].join(' ');
            /**
             * Reject action.
             *
             * @param message error message
             * @param payload R
             * @returns ErrorAction<typeof rejectedType, F>
             */
            function rejected(message, payload) {
                return {
                    type: rejectedType,
                    payload,
                    error: {
                        message
                    },
                    showMessage: true
                };
            }
            rejected.type = rejectedType;
            rejected.match = createMatcher(rejectedType);
            return {
                pending,
                fulfilled,
                rejected
            };
        };
    }
    const EXTERNAL_ACTION_PREFIX = '[ext]';
    const createExternalAction = createActionFactory(EXTERNAL_ACTION_PREFIX);
    const iconsLoaded = createExternalAction('icons-loaded');
    const controlSelected = createExternalAction('control-selected');
    const selectControl = createExternalAction('select-control');
    const addExtensionPoint = createExternalAction('add-extension-point');
    const deletePropertyChanges = createExternalAction('delete-property-changes');
    const outlineChanged = createExternalAction('outline-changed');
    const changeProperty = createExternalAction('change-property');
    const propertyChanged = createExternalAction('property-changed');
    const propertyChangeFailed = createExternalAction('change-property-failed');
    const changeStackModified = createExternalAction('change-stack-modified');
    const toggleAppPreviewVisibility = createExternalAction('toggle-app-preview-visibility');
    const reloadApplication = createExternalAction('reload-application');
    const storageFileChanged = createExternalAction('storage-file-changed');
    const setAppMode = createExternalAction('set-app-mode');
    const applicationModeChanged = createExternalAction('application-mode-changed');
    const setUndoRedoEnablement = createExternalAction('set-undo-redo-enablement');
    const setSaveEnablement = createExternalAction('set-save-enablement');
    const appLoaded = createExternalAction('app-loaded');
    const undo = createExternalAction('undo');
    const redo = createExternalAction('redo');
    const save = createExternalAction('save');
    const quickActionListChanged = createExternalAction('quick-action-list-changed');
    const updateQuickAction = createExternalAction('update-quick-action');
    const executeQuickAction = createExternalAction('execute-quick-action');
    const executeContextMenuAction = createExternalAction('execute-context-menu-action');
    const setApplicationRequiresReload = createExternalAction('set-application-requires-reload');
    const externalFileChange = createExternalAction('external-file-change');
    const showInfoCenterMessage = createExternalAction('show-info-center-message');
    const createAsyncExternalAction = createAsyncActionFactory(EXTERNAL_ACTION_PREFIX);
    const requestControlContextMenu = createAsyncExternalAction('request-control-context-menu');

    /**
     * Returns a function which calls the callback function only after the specified idle time
     * Works similar to the Debounce operator from rxjs https://reactivex.io/documentation/operators/debounce.html link.
     *
     * @param callback Function to execute
     * @param delay Idle period in milliseconds after which the callback will be executed
     * @returns A wrapper function that should be called to invoke the callback function after delay
     */
    function debounce(callback, delay) {
        let timerId;
        return (...args) => {
            clearTimeout(timerId);
            timerId = setTimeout(() => {
                callback(...args);
            }, delay);
        };
    }

    const POST_MESSAGE_ACTION_TYPE = 'post-message-action';
    /**
     * Check if data isPostMessageAction.
     *
     * @param data - post message action
     * @returns data is PostMessageAction<T>
     */
    function isPostMessageAction(data) {
        return data?.type === POST_MESSAGE_ACTION_TYPE && typeof data?.action === 'object';
    }
    /**
     * Method to start post message communication.
     *
     * @param target target window
     * @param onActionHandler action handler
     * @param logger to log info, default: browser console.
     * @returns PostMessageCommunication<T>
     */
    function startPostMessageCommunication(target, onActionHandler, logger = console) {
        /**
         * Returns target windows or undefined.
         *
         * @returns Window | undefined
         */
        function getTarget() {
            if (typeof target === 'function') {
                return target();
            }
            return target;
        }
        /**
         * Invoke action on post message.
         *
         * @param event event
         */
        function postMessageListener(event) {
            const target = getTarget();
            if (event.origin !== target?.origin || event.source !== target) {
                // Ignore messages from unknown sources
                return;
            }
            if (isPostMessageAction(event.data)) {
                onActionHandler(event.data.action).catch((error) => logger.error(error));
            }
            else {
                logger.warn(`Unknown message received ${event.data}`);
            }
        }
        function dispose() {
            window.removeEventListener('message', postMessageListener);
        }
        /**
         * Post message to a give window.
         *
         * @param action action payload
         */
        function sendAction(action) {
            const target = getTarget();
            if (!target) {
                return;
            }
            const message = {
                type: POST_MESSAGE_ACTION_TYPE,
                action
            };
            target.postMessage(message, target.origin);
        }
        window.addEventListener('message', postMessageListener);
        return {
            dispose,
            sendAction
        };
    }

    let enabled = false;
    function enableTelemetry() {
        enabled = true;
    }
    function disableTelemetry() {
        enabled = false;
    }
    /**
     * Reports telemetry data from Control Property Editor.
     *
     * @param data The TelemetryData object, that needs to be reported
     * @returns {Promise<void>}
     */
    async function reportTelemetry(data) {
        try {
            if (enabled) {
                const requestOptions = {
                    method: 'POST',
                    headers: {
                        Accept: 'application/json',
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(data)
                };
                await fetch('/preview/api/telemetry', requestOptions);
            }
        }
        catch (_error) {
            // something is wrong with the telemetry service
            disableTelemetry();
        }
    }

    const isUpperCase = (code) => code >= 65 && code <= 90;
    const isLowerCase = (code) => code >= 97 && code <= 122;
    const isWordDelimiter = (code) => code === 45 || code === 95 || code === 32; // - or _ or space
    const convertCamelCaseToPascalCase = (text) => {
        const words = [];
        let word = '';
        let lookForUpperCase = true;
        for (let i = 0; i < (text ?? '').length; i++) {
            const character = text[i];
            if (lookForUpperCase) {
                // make sure that the first letter is capitalized
                word += word.length === 0 ? character.toUpperCase() : character;
                if (isLowerCase(text.charCodeAt(i + 1))) {
                    // First lower case character after upper case character -> switch mode to collect only lower case characters
                    lookForUpperCase = false;
                }
                else if (isUpperCase(text.charCodeAt(i + 1)) && isLowerCase(text.charCodeAt(i + 2))) {
                    // Next character is the last uppercase character after a sequence of upper case character -> create an abbreviated word
                    words.push(word);
                    word = '';
                }
            }
            else if (isUpperCase(text.charCodeAt(i))) {
                // Upper case character indicates the beginning of a new word -> switch mode to detect abbreviated word
                words.push(word);
                lookForUpperCase = true;
                word = character;
            }
            else if (isWordDelimiter(text.charCodeAt(i))) {
                words.push(word);
                lookForUpperCase = true;
                word = '';
            }
            else {
                word += character;
            }
        }
        if (word.length) {
            words.push(word);
        }
        return words.join(' ');
    };
    exports.FlexChangesEndPoints = void 0;
    (function (FlexChangesEndPoints) {
        FlexChangesEndPoints["changes"] = "/preview/api/changes";
    })(exports.FlexChangesEndPoints || (exports.FlexChangesEndPoints = {}));

    const __esModule = true ;

    exports.BOOLEAN_VALUE_TYPE = BOOLEAN_VALUE_TYPE;
    exports.CHECKBOX_EDITOR_TYPE = CHECKBOX_EDITOR_TYPE;
    exports.CONTROL_CHANGE_KIND = CONTROL_CHANGE_KIND;
    exports.DROPDOWN_EDITOR_TYPE = DROPDOWN_EDITOR_TYPE;
    exports.EXTERNAL_ACTION_PREFIX = EXTERNAL_ACTION_PREFIX;
    exports.FLOAT_VALUE_TYPE = FLOAT_VALUE_TYPE;
    exports.FULFILLED_SUFFIX = FULFILLED_SUFFIX;
    exports.GENERIC_CHANGE_KIND = GENERIC_CHANGE_KIND;
    exports.INPUT_EDITOR_TYPE = INPUT_EDITOR_TYPE;
    exports.INTEGER_VALUE_TYPE = INTEGER_VALUE_TYPE;
    exports.NESTED_QUICK_ACTION_KIND = NESTED_QUICK_ACTION_KIND;
    exports.PENDING_CHANGE_TYPE = PENDING_CHANGE_TYPE;
    exports.PENDING_SUFFIX = PENDING_SUFFIX;
    exports.REJECTED_SUFFIX = REJECTED_SUFFIX;
    exports.SAVED_CHANGE_TYPE = SAVED_CHANGE_TYPE;
    exports.SCENARIO = SCENARIO;
    exports.SIMPLE_QUICK_ACTION_KIND = SIMPLE_QUICK_ACTION_KIND;
    exports.STRING_VALUE_TYPE = STRING_VALUE_TYPE;
    exports.UNKNOWN_CHANGE_KIND = UNKNOWN_CHANGE_KIND;
    exports.__esModule = __esModule;
    exports.addExtensionPoint = addExtensionPoint;
    exports.appLoaded = appLoaded;
    exports.applicationModeChanged = applicationModeChanged;
    exports.changeProperty = changeProperty;
    exports.changeStackModified = changeStackModified;
    exports.controlSelected = controlSelected;
    exports.convertCamelCaseToPascalCase = convertCamelCaseToPascalCase;
    exports.createAsyncActionFactory = createAsyncActionFactory;
    exports.debounce = debounce;
    exports.deletePropertyChanges = deletePropertyChanges;
    exports.disableTelemetry = disableTelemetry;
    exports.enableTelemetry = enableTelemetry;
    exports.executeContextMenuAction = executeContextMenuAction;
    exports.executeQuickAction = executeQuickAction;
    exports.externalFileChange = externalFileChange;
    exports.iconsLoaded = iconsLoaded;
    exports.outlineChanged = outlineChanged;
    exports.propertyChangeFailed = propertyChangeFailed;
    exports.propertyChanged = propertyChanged;
    exports.quickActionListChanged = quickActionListChanged;
    exports.redo = redo;
    exports.reloadApplication = reloadApplication;
    exports.reportTelemetry = reportTelemetry;
    exports.requestControlContextMenu = requestControlContextMenu;
    exports.save = save;
    exports.selectControl = selectControl;
    exports.setAppMode = setAppMode;
    exports.setApplicationRequiresReload = setApplicationRequiresReload;
    exports.setSaveEnablement = setSaveEnablement;
    exports.setUndoRedoEnablement = setUndoRedoEnablement;
    exports.showInfoCenterMessage = showInfoCenterMessage;
    exports.startPostMessageCommunication = startPostMessageCommunication;
    exports.storageFileChanged = storageFileChanged;
    exports.toggleAppPreviewVisibility = toggleAppPreviewVisibility;
    exports.undo = undo;
    exports.updateQuickAction = updateQuickAction;

}));
